File: day_08/day8.jl
#=
day8:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-08

Added some tests to the code so I can check that I haven't broken anything
while working on it/refactoring etc.
=#

import Test

mutable struct Forest
    height::Int
    width::Int
    trees::Array{Int, 2}

    function Forest(data::Array{String})
        this = new()
        this.height = length(data)
        this.width = length(data[1])
        arrayofarrays = [row2ints(row) for row in data]
        # Convert an array of arrays into a Matrix
        # Don't underatand this yet. Found it on Stackoverflow
        this.trees = vcat(transpose.(arrayofarrays)...)
        #=
        What it seems to do:
        transpose.(a) the . calls transpose on each 'row' in the array instead of just once on the
        whole array.
        vcat( ...) the ... splits the output of transpose into elements and passes the to vcat
        which combines them into a matrix.
        =#
        return this
    end
end

row2ints(line) = [parse(Int, t) for t in split(line,"")]


function count(forest::Forest)
    count = 0
    for y in range(1,forest.height)
        for x in range(1,forest.width)
            count += isvisible(forest, x, y)
        end
    end
    return count
end

function count_in_direction(tree, trees, direction)
    l = length(trees)
    if direction == 1
        for i in range(1,l)
            if trees[i] >= tree
                return i
            end
        end
    end
    if direction == -1
        for i in range(1,l)
            if trees[1 + l - i] >= tree
                return i
            end
        end
    end
    return length(trees)
end

function scenic_score(forest::Forest, x::Int, y::Int)
    row = forest.trees[y, :]
    column = forest.trees[:, x]
    tree = forest.trees[y, x]
    return (
            count_in_direction(tree, row[x + 1:end], 1) *
            count_in_direction(tree, row[1:x-1], -1) *
            count_in_direction(tree, column[y + 1:end], 1) *
            count_in_direction(tree, column[1:y-1], -1)
    )
end

function check_row(x::Int, row)
    tree = row[x]
    return (all([t < tree for t in row[1:x - 1]]) || all([t < tree for t in row[x + 1:end]]))
end

function check_column(y::Int, column)
    tree = column[y]
    return (all([t < tree for t in column[1:y - 1]]) || all([t < tree for t in column[y + 1:end]]))
end

function isvisible(forest::Forest, x::Int, y::Int)
    row = forest.trees[y, :]
    column = forest.trees[:, x]
    if x == 1 ||  y == 1 || x == forest.width || y == forest.height ||
            check_row(x, row) || check_column(y, column)
        return 1
    else
        return 0
    end
end

function max_scenic(forest::Forest)
    score = 0
    for y in range(1, forest.height)
        for x in range(1, forest.width)
            score = max(score, scenic_score(forest, x, y))
        end
    end
    return score
end

# The test data set
testdata = readlines(open("day_08/test.txt"))
forest = Forest(testdata)

Test.@test count_in_direction(5, [5, 3, 3], -1) == 3
Test.@test count_in_direction(5, [1, 5, 3], 1) == 2
Test.@test count(forest) == 21
Test.@test scenic_score(forest, 2, 3) == 6
Test.@test max_scenic(forest) == 8


# The test data
forest = Forest(readlines(open("day_08/data.txt")))

# Part 1
println("Part 1")
println(count(forest))


# Part 2
println("Part 2")
println(max_scenic(forest))

File: day_08/day8.py
from aoc import aoc
import numpy as np
from time import perf_counter


class Forest:
    def __init__(self, grid):
        self.heights = np.array([[int(t) for t in row] for row in grid])
        self.height, self.width = len(grid), len(grid[0])

    def isvisible(self, x, y):
        if x == 0 or y == 0 or x == (self.width - 1) or y == (self.height - 1):
            return 1
        row = self.heights[y]
        tree = row[x]
        # This works because we're using numpy arrays not python lists.
        if all(row[:x] < tree) or all(row[x + 1:] < tree):
            return 1
        column = self.heights[:, x]
        if all(column[:y] < tree) or all(column[y + 1:] < tree):
            return 1
        return 0

    def scenic_score(self, x, y):
        row = self.heights[y]
        column = self.heights[:, x]
        tree = self.heights[y][x]
        return (
                count_in_dir(tree, row[x + 1:], 1) *
                count_in_dir(tree, row[:x], -1) *
                count_in_dir(tree, column[y + 1:], 1) *
                count_in_dir(tree, column[:y], -1)
        )

    def count(self):
        count = 0
        for y in range(self.height):
            for x in range(self.width):
                count += self.isvisible(x, y)
        return count

    def max_scenic(self):
        score = 0
        for y in range(self.height):
            for x in range(self.width):
                score = max(score, self.scenic_score(x, y))
        return score


def count_in_dir(tree, trees, direction):
    l = len(trees)
    if direction == 1:
        for i in range(l):
            if trees[i] >= tree:
                return 1 + i
    if direction == -1:
        for i in range(l):
            if trees[l - i - 1] >= tree:
                return i + 1
    return len(trees)


def part1(forest):
    count = forest.count()
    assert count == 21 or count == 1717
    return count


def part2(forest):
    score = forest.max_scenic()
    assert score == 8 or score == 321975
    return score


def run(file):
    print("Part 1")
    data = aoc.loadlines(file)
    forest = Forest(data)
    print(part1(forest))
    print("Part 2")
    print(part2(forest))


if __name__ == "__main__":
    start = perf_counter()
    print("Test Data")
    run("test.txt")
    print("Actual Data")
    run("data.txt")
    end = perf_counter()
    print(end - start)
